frb_rust/src/pool.rs

```diff
@@ -0,0 +1,235 @@
+//! Original: wasm_bindgen raytrace-parallel example
+//!
+//! File: https://github.com/rustwasm/wasm-bindgen/blob/main/examples/raytrace-parallel/src/pool.rs
+
+use crate::TransferClosure;
+use crate::{console_error, console_log};
+use js_sys::Array;
+use std::cell::RefCell;
+use std::iter::FromIterator;
+use std::rc::Rc;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::BlobPropertyBag;
+use web_sys::{Blob, Url};
+use web_sys::{DedicatedWorkerGlobalScope, MessageEvent};
+use web_sys::{ErrorEvent, Event, Worker};
+
+#[wasm_bindgen]
+pub struct WorkerPool {
+    state: Rc<PoolState>,
+    name: String,
+}
+
+struct PoolState {
+    workers: RefCell<Vec<Worker>>,
+    callback: Closure<dyn FnMut(Event)>,
+}
+
+#[wasm_bindgen]
+impl WorkerPool {
+    /// Creates a new `WorkerPool` which immediately creates `initial` workers.
+    ///
+    /// The pool created here can be used over a long period of time, and it
+    /// will be initially primed with `initial` workers. Currently workers are
+    /// never released or gc'd until the whole pool is destroyed.
+    ///
+    /// # Errors
+    ///
+    /// Returns any error that may happen while a JS web worker is created and a
+    /// message is sent to it.
+    #[wasm_bindgen(constructor)]
+    pub fn new(initial: usize, name: String) -> Result<WorkerPool, JsValue> {
+        let pool = WorkerPool {
+            name,
+            state: Rc::new(PoolState {
+                workers: RefCell::new(Vec::with_capacity(initial)),
+                callback: Closure::new(|event: Event| {
+                    console_log!("unhandled event: {}", event.type_());
+                }),
+            }),
+        };
+        for _ in 0..initial {
+            let worker = pool.spawn()?;
+            pool.state.push(worker);
+        }
+
+        Ok(pool)
+    }
+
+    /// Unconditionally spawns a new worker
+    ///
+    /// The worker isn't registered with this `WorkerPool` but is capable of
+    /// executing work for this wasm module.
+    ///
+    /// # Errors
+    ///
+    /// Returns any error that may happen while a JS web worker is created and a
+    /// message is sent to it.
+    fn spawn(&self) -> Result<Worker, JsValue> {
+        // console_log!("spawning new worker");
+        // TODO: what do do about `./worker.js`:
+        //
+        // * the path is only known by the bundler. How can we, as a
+        //   library, know what's going on?
```

Comment:
This was a comment from the original repo, but the concern applies to us as well.